storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serialization/RemoteLogMetadataSerde.java

```diff
@@ -0,0 +1,105 @@
+import org.apache.kafka.server.log.remote.metadata.storage.generated.RemotePartitionDeleteMetadataRecord;
+import org.apache.kafka.server.log.remote.storage.RemoteLogMetadata;
+import org.apache.kafka.server.log.remote.storage.RemoteLogSegmentMetadata;
+import org.apache.kafka.server.log.remote.storage.RemoteLogSegmentMetadataUpdate;
+import org.apache.kafka.server.log.remote.storage.RemotePartitionDeleteMetadata;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * This class provides serialization and deserialization for {@link RemoteLogMetadata}. This is the root serde
+ * for the messages that are stored in internal remote log metadata topic.
+ */
+public class RemoteLogMetadataSerde {
+    private static final short REMOTE_LOG_SEGMENT_METADATA_API_KEY = new RemoteLogSegmentMetadataRecord().apiKey();
+    private static final short REMOTE_LOG_SEGMENT_METADATA_UPDATE_API_KEY = new RemoteLogSegmentMetadataUpdateRecord().apiKey();
+    private static final short REMOTE_PARTITION_DELETE_API_KEY = new RemotePartitionDeleteMetadataRecord().apiKey();
+
+    private final Map<String, Short> remoteLogStorageClassToApiKey;
+    private final Map<Short, RemoteLogMetadataTransform> keyToTransform;
+    private final BytesApiMessageSerde bytesApiMessageSerde;
+
+    public RemoteLogMetadataSerde() {
+        remoteLogStorageClassToApiKey = createRemoteLogStorageClassToApiKeyMap();
+        keyToTransform = createRemoteLogMetadataTransforms();
+        bytesApiMessageSerde = new BytesApiMessageSerde() {
+            @Override
+            public ApiMessage apiMessageFor(short apiKey) {
+                return newApiMessage(apiKey);
+            }
+        };
+    }
+
+    protected ApiMessage newApiMessage(short apiKey) {
+        return MetadataRecordType.fromId(apiKey).newMetadataRecord();
+    }
+
+    protected Map<Short, RemoteLogMetadataTransform> createRemoteLogMetadataTransforms() {
+        Map<Short, RemoteLogMetadataTransform> map = new HashMap<>();
+        map.put(REMOTE_LOG_SEGMENT_METADATA_API_KEY, new RemoteLogSegmentMetadataTransform());
+        map.put(REMOTE_LOG_SEGMENT_METADATA_UPDATE_API_KEY, new RemoteLogSegmentMetadataUpdateTransform());
+        map.put(REMOTE_PARTITION_DELETE_API_KEY, new RemotePartitionDeleteMetadataTransform());
+        return map;
+    }
+
+    protected Map<String, Short> createRemoteLogStorageClassToApiKeyMap() {
+        Map<String, Short> map = new HashMap<>();
+        map.put(RemoteLogSegmentMetadata.class.getName(), REMOTE_LOG_SEGMENT_METADATA_API_KEY);
+        map.put(RemoteLogSegmentMetadataUpdate.class.getName(), REMOTE_LOG_SEGMENT_METADATA_UPDATE_API_KEY);
+        map.put(RemotePartitionDeleteMetadata.class.getName(), REMOTE_PARTITION_DELETE_API_KEY);
+        return map;
+    }
+
+    public byte[] serialize(RemoteLogMetadata remoteLogMetadata) {
+        Short apiKey = remoteLogStorageClassToApiKey.get(remoteLogMetadata.getClass().getName());
```

Comment:
This 2-way map lookup feels a bit complex to me. It appears here the requirement is that you need the `apiKey` corresponding to the provided `RemoteLogMetadata`. To make it simpler, why not provide a `short apiKey()` abstract method in the `RemoteLogMetadata` base class and then ask the specializing classes to implement it? You will use then use the `remoteLogMetadata.apiKey()` method to get the apiKey here. This will avoid the need to maintain 2 maps within this class, you will only need the `remoteLogMetadataTransform` map.

Reply:
`apiKey` is an internal serialization detail of one of the implementations. We should not leak that to public API like `RemoteLogMetadata` and its subclasses.

Reply:
I guess we can continue discussion here: https://github.com/apache/kafka/pull/10271#discussion_r622786068 ?